[JEP-222] Add experimental WebSocket support + Jetty update in Winstone + Remoting 4.0

jglick opened the pull request:
See [JEP-222](https://jenkins.io/jep/222). Downstream of ~https://github.com/jenkinsci/winstone/pull/79~ + ~https://github.com/jenkinsci/remoting/pull/357~ + ~https://github.com/jenkinsci/jenkins-test-harness/pull/183~.

### Proposed Changelog Entries

* Added **experimental** support for connecting inbound agents using a WebSocket transport over HTTP(S) rather than the TCP port, for easier network setup. See [JEP-222](https://jenkins.io/jep/222) for details.
* Remoting 4.0 release, mostly to highlight the binary incompatibility in the last 3.x release: https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0
* Jetty update in Winstone: https://github.com/jenkinsci/winstone/releases/tag/winstone-5.5

jglick committed:
Playing with WebSocket connections.

jglick committed:
Pluggable handlers.

jglick committed:
errorWithoutStack

jglick committed:
Server-side keepalive pings.

jglick committed:
Redesigned to work as an HttpResponse.

jglick committed:
Unused dep.

jglick committed:
Comment on JnlpSlaveRestarterInstaller.

jglick committed:
Merge branch 'master' into websocket

jglick committed:
Sketch of WebSocket-based endpoint for JNLPLauncher.

jglick committed:
Comment.

jglick committed:
Merge branch 'master' into websocket

jglick committed:
Merge branch 'master' into websocket

jglick committed:
Test enhancement.

jglick committed:
Making it more obvious from JNLPLauncherTest where remoting.jar is being loaded from.

jglick committed:
Reworked protocol to negotiate remote capabilities.

jglick committed:
Unhelpful comment.

jglick committed:
Simplifying handshake to use HTTP headers.

jglick committed:
Moving code into top-level classes and otherwise prettifying.

jglick committed:
Linking to upstream PRs.

jglick committed:
Timestamped snapshot + incremental.

jglick committed:
Merge branch 'master' into websocket

jglick committed:
Picking up incrementalified build of winstone.

jglick committed:
Working around https://github.com/kohsuke/access-modifier/pull/17.

jglick commented:
Looks like there are some test failures.

jglick committed:
Merge branch 'master' into websocket

jglick committed:
Finally have an incrementalified build of remoting.

jeffret-b commented on core/src/main/java/jenkins/slaves/WebSocketAgents.java:
```diff
@@ -0,0 +1,187 @@
+
+        @Override
+        protected boolean keepAlive() {
+            return true; // Remoting ping thread may be too slow
```
Should the Remoting ping thread be adjusted to be faster in these scenarios? Or turned off in favor of other mechanisms, like this? Duplicate ping mechanisms over a channel can sometimes get in each other's way.

I see some later comments about necessary ping intervals based on different technologies. I'm not sure what makes the most sense here.

jglick replied:
I think it is OK to keep both, at least pending further experience. The Remoting ping is by default every 5m with a 4m timeout, and beyond network connectivity would verify that both sides of the channel are actually processing packets. The WebSocket ping introduced here is unidirectional (server → client), does not wait for pongs unless the underlying library does it automatically, and serves only to prevent certain reverse proxies from closing the connection prematurely.

jeffret-b commented on core/src/main/java/jenkins/websocket/WebSocketEcho.java:
```diff
@@ -0,0 +1,69 @@
+
+@Extension
+@Restricted(NoExternalUse.class)
+public class WebSocketEcho  extends InvisibleAction implements UnprotectedRootAction {
```
I'm not understanding yet why this is needed.

jglick replied:
This is just here for testing and diagnostics—it is the most convenient way to check whether Jenkins is actually serving WebSocket connections, including through load balancers and whatnot, and to verify keepalive behavior. [JEP-222: _Interactive Tests_](https://github.com/jenkinsci/jep/blob/master/jep/222/README.adoc#interactive-tests)

jeffret-b replied:
I'm a little concerned about potential security implications here. Not that I can enumerate any specific ones, though. It's just that this provides an unauthenticated channel. There probably isn't a a DOS attack, because there isn't any noticeable multiplier. Hmmm ... maybe it could be used to pump a lot of worthless, unauthenticated data through a Jenkins server. That sounds like a DOS attack.

jglick replied:
More than just hitting, say, `http://jenkins/login` repeatedly?

jeffret-b replied:
I don't know. It seems like this allows someone to open a socket and just stream data through and back, which seems to be worse than just hitting the login page. Maybe it should just send back an abbreviated or limited length message. The first 10 characters or something.

jglick replied:
I cannot see that making any difference; one way or another it needs to handle the incoming message.

Maybe I will just restrict this to administrators. (Unlike TCP connections, it is trivial to pass authentication.)

jeffret-b replied:
True. I do like the idea of restricting it to administrators. That removes concerns.

jglick replied:
[Serendipitously](https://github.com/vi/websocat/releases/tag/v1.5.0)

> Using client URI's like `websocat ws://user:password@host/` now adds basic authentication HTTP header to request

jeffret-b replied:
That's nicely handy and timely. The behavior seems reasonable and works fine now.

jeffret-b reviewed (commented):
So far I've just done a quick overview review. It looks good. It's well structured and easy to follow. I like how nicely it fits into the existing structure while providing valuable new capability.

I'll do a more thorough review at a later time.

jglick committed:
Merge branch 'master' into websocket

jglick commented on test-pom/pom.xml:
```diff
@@ -71,7 +71,7 @@ THE SOFTWARE.
       <groupId>${project.groupId}</groupId>
       <artifactId>jenkins-test-harness</artifactId>
-      <version>2.55</version>
+      <version>2.57-rc1258.021b941d1c70</version> <!-- TODO https://github.com/jenkinsci/jenkins-test-harness/pull/183 -->
```
note https://github.com/apache/netbeans/pull/1766

jglick committed:
Merge branch 'jth-jetty-upgrade' into websocket

jglick committed:
Use -webSocket option.

jglick committed:
Handling some close and error methods.

jglick committed:
Sending X-Remoting-Minimum-Version.

jglick committed:
If hudson.remoting.jnlp.Main._main fails, show the error.

jglick committed:
Comment.

jglick committed:
GUI analogue of https://github.com/jenkinsci/remoting/pull/357/commits/86cea5b83fd2fc93a9fc10896962eb42c95f80a0.

jglick committed:
Missing since tags.

jglick committed:
Capitalization.

jglick committed:
WebSockets.isSupported

jglick committed:
Rather than hiding JNLPLauncher.DescriptorImpl when the TCP port is disabled, always display it, but show form validation appropriate to WebSocket or TCP mode.

jglick committed:
Form validation fixes.

jglick committed:
Minor test improvements.

jglick committed:
Reworked WebSocketAgents to be compatible with JnlpAgentReceiver.

jtnord commented on core/src/main/java/hudson/slaves/JNLPLauncher.java:
```diff
@@ -143,6 +147,21 @@ public boolean isLaunchSupported() {
+    /**
+     * @since TODO
+     */
+    public boolean isWebSocket() {
```
should this be a boolean.  with kafka for remoting and others I am wondering if it should be an enum or something else.

jglick replied:
The Kafka implementation as I understand it is quite different and does not integrate with `JNLPLauncher`.

See also [discussion](https://github.com/jenkinsci/jep/blob/master/jep/222/README.adoc#dedicated-computerlauncher).

jtnord commented on core/src/main/java/jenkins/slaves/WebSocketAgents.java:
```diff
@@ -0,0 +1,193 @@
+ * THE SOFTWARE.
+ */
+
+package jenkins.slaves;
```
probably we can't but could we put new code in `jenkins.agents`?

jglick replied:
The code base is already split between `hudson.slaves` and `jenkins.slaves`, and various API types already use the term _slave_, so this feels impractical.

jglick replied:
As you prefer.

jtnord commented on test/src/test/java/hudson/slaves/JNLPLauncherTest.java:
```diff
@@ -69,7 +71,9 @@
-    @Rule public TemporaryFolder tmpDir = new TemporaryFolder(); 
+    @Rule public TemporaryFolder tmpDir = new TemporaryFolder();
+
+    @Rule public LoggerRule logging = new LoggerRule().record(Slave.class, Level.FINE);
```
unused in test - is this to show that your fine logging does not blow up due to invalid parameters or other. (in which case a comment in the code would help future reviewers)

jglick replied:
Does not affect test pass status, just there to help developers understand what is happening when the test is being run interactively. I.e., the typical usage of `LoggerRule`.

jglick committed:
Removing WebSocketSession.keepAlive in favor of a global setting applicable to all services.

jglick committed:
Add a -webSocket option to the Jenkins CLI.

jglick committed:
Using a snapshot deployment of Remoting, pending INFRA-2379.

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
After #3838 there is no reason to recheck authentication after parsing CLICommand arguments.

jglick committed:
Advertise the -webSocket option.

jglick committed:
Adapt to newer HtmlUnit.

jglick committed:
Merge branch 'dns' into websocket

jglick committed:
Merge branch 'htmlunit' into websocket

jglick committed:
Merge branch 'master' into htmlunit

jglick committed:
https://github.com/HtmlUnit/htmlunit/pull/29 seems to have been incompatible.
WebClient.addRequestHeader will no longer override a header in a WebRequest,
such as when the same WebClient/WebRequest was previously used with different headers.

jglick committed:
Tracked down a behavioral change in passing through URL-encoded path characters.
https://github.com/HtmlUnit/htmlunit/commit/2c4956863420e4baef9d3d8c23ec0577ec64d2bf picks up
https://github.com/apache/httpcomponents-client/commit/8c04c6ae5e5ba1432e40684428338ce68431766b which is the actual change.

jglick committed:
Merge branch 'htmlunit' into websocket

jglick commented on cli/pom.xml:
```diff
@@ -93,6 +93,12 @@
+      <groupId>org.glassfish.tyrus.bundles</groupId>
+      <artifactId>tyrus-standalone-client-jdk</artifactId>
+      <version>1.12</version>
+      <optional>true</optional> <!-- ditto -->
```
https://github.com/jenkinsci/remoting/pull/357#discussion_r360519231

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
Disabled TCP port does not matter in WebSocket mode.

jglick committed:
Shade dependencies needed for jenkins-cli.jar.

jglick committed:
Help text edit.

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket
Resolves conflict with #4378.

jglick committed:
https://github.com/jenkinsci/winstone/pull/79 was released as 5.5.

jglick committed:
https://github.com/jenkinsci/jenkins-test-harness/pull/183 released as 2.59.

jglick committed:
Merge branch 'removeDeprecatedProtocols' of github.com:jeffret-b/jenkins into websocket

jglick committed:
https://github.com/jenkinsci/jenkins/pull/4387#discussion_r362696234

jglick committed:
Bumping remoting to a new deployed snapshot.

jeffret-b commented:
I pulled down this PR branch and built it but when I try to run I get
```
Exception in thread "main" java.lang.IllegalAccessException: Can not set static final java.lang.String field winstone.WinstoneSession.SESSION_COOKIE_NAME to java.lang.String
```
Is there anything else I need to update or configure to get this latest version to run?

jglick commented:
I have not seen such an error. Are you using Java 11?

jeffret-b commented:
> I have not seen such an error. Are you using Java 11?

No, I'm on Java 8. It worked just fine yesterday. I pulled down the latest this morning and rebuilt and now it isn't working. I don't have a clue why yet.

jglick commented:
I am on it.

jglick committed:
Need https://github.com/jenkinsci/winstone/pull/86.

jglick committed:
https://github.com/jenkinsci/winstone/pull/86 released as 5.6.

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
Introduced constant for X-Remoting-Minimum-Version.

jglick committed:
s/slave/agent/ in GUI

jglick committed:
Removing in-JVM test, as it was no longer useful after introducing shading anyway.

jeffret-b reviewed (commented):
The code looks good. I left a few comments.

I want to test this out before stamping it with an approval. I left a few notes to myself of what to try. I'll try to do the testing tomorrow.

jglick commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -101,8 +113,9 @@ public static int _main(String[] _args) throws Exception {
-        boolean tryLoadPKey = true;
+        boolean noKeyAuth = false;
 
+        // TODO perhaps allow mode to be defined by environment variable too (assuming $JENKINS_USER_ID can be used for -user)
```
I just thought it would be convenient—while you can set the Jenkins URL in a profile (or `withEnv`…), you cannot currently do the same for the mode.

jeffret-b replied:
Maybe. If you want to do, I won't object, but I'm just not convinced of the value.

jglick replied:
Not intending to do it here; just left as a possibility for the future.

jglick commented on core/src/main/java/hudson/slaves/JNLPLauncher.java:
```diff
@@ -196,25 +215,23 @@ public boolean isWorkDirSupported() {
+                if (!WebSockets.isSupported()) {
+                    return FormValidation.error("WebSocket support is not enabled in this Jenkins installation");
+                }
+                if (Util.fixEmptyAndTrim(tunnel) != null) {
```
Right, it is mutually exclusive in `remoting.jar`, I am just doing a precheck here so you do not get as far as trying to actually launch the agent and then seeing an error from the CLI.

jglick committed:
Bump.

jglick commented on core/src/main/java/jenkins/slaves/WebSocketAgents.java:
```diff
@@ -0,0 +1,189 @@
+            LOGGER.warning(() -> "no such agent " + agent);
+            throw HttpResponses.errorWithoutStack(400, "no such agent");
+        }
+        if (!MessageDigest.isEqual(secret.getBytes(StandardCharsets.US_ASCII), JnlpAgentReceiver.DATABASE.getSecretOf(agent).getBytes(StandardCharsets.US_ASCII))) { // TODO unless anonymous has CONNECT?
```
There is special handling in the GUI to decide whether or not to show the JNLP button, but I agree nothing is needed here.

jglick committed:
No need to check for anonymous CONNECT here.

jglick committed:
s/jenkins.slaves/jenkins.agents/g for new code.

jglick committed:
Restrict the diagnostic endpoint to administrators.

jglick committed:
Fixed Javadoc import after package move.

jeffret-b commented on core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly:
```diff
@@ -28,6 +28,9 @@ THE SOFTWARE.
   <j:if test="${descriptor.workDirSupported}"> 
     <f:property title="${%Enable work directory}" field="workDirSettings" />
   </j:if>
+  <f:entry field="webSocket">
```
This works a little differently than I expected. Or a little inconsistently.

A) If I don't enable WebSocket here:
  1) I can launch an agent normally and it connects via the TCP port.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

B) If I do enable WebSocket here:
  1) If I try to launch an agent normally, it gets the "-webSocket" argument in the jnlp file and connects via WebSocket.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

In other words, the agent can control if it wants to connect via WebSocket or direct but it can't control if it wants to connect via jnlp file to TCP port.

Things are better when they are regular and consistent. But I'm not certain this is really a problem here, that it needs to be changed, or what it would be changed to. 

Theoretically the config UI could have a choice to connect via i) config (jnlp) file, ii) direct, or iii) WebSocket and then the server would only allow that mechanism. 

Another possibility would be that the WebSocket implementation check connections and not allow them if they're not configured for WebSocket.

The inconsistent behavior had me confused for a while until I mapped it out but maybe it's not going to be confusing for users and developers.

Just recording my observations and musing about it.

jglick replied:
You are correct that the server does not attempt to reject connections made using a different transport than the selected one. (`-direct` is not a distinct option—this is still using the TCP port, just with the port number being discovered differently.) The checkbox controls whether the CLI argument `-webSocket` is offered to `Main` when `Launcher` is run with `-jnlpUrl`. In other words, the server merely uses this to serve up appropriate `/computer/xxx/slave-agent.jnlp` content.

jglick replied:
Remember that `Main` is the fundamental entry point for inbound agents. `Launcher` is the wrapper entry point that parses JNLP to call `Main`, as well as handling stdio mode for outbound agents (and a TCP _server_ mode which AFAIK is unused by any Jenkins `ComputerLauncher`). It would be nice if Javadoc and/or Markdown clarified the distinction, as it is hard to keep track of (see https://github.com/jenkinsci/docker-jnlp-slave/issues/87 for example though I think I am abandoning that).

jeffret-b replied:
Yeah, I think the existing behavior is pretty reasonable the more I think about it. I wanted to record my observations and see if anything else popped out.

jeffret-b commented on cli/src/main/java/hudson/cli/CLI.java:
```diff
@@ -91,7 +103,7 @@ public static void main(final String[] _args) throws Exception {
     }
 
-    private enum Mode {HTTP, SSH}
+    private enum Mode {HTTP, SSH, WEB_SOCKET}
```
Works fine for the things I tried.

jeffret-b commented on core/src/main/resources/hudson/slaves/SlaveComputer/slave-agent.jnlp.jelly:
```diff
@@ -63,6 +63,9 @@ THE SOFTWARE.
       <application-desc main-class="hudson.remoting.jnlp.Main">
         <argument>${it.jnlpMac}</argument>
         <argument>${it.node.nodeName}</argument>
+        <j:if test="${launcher.webSocket}">
```
Works fine for the basic scenarios I've tested.

jeffret-b reviewed (approved):
This looks nice. Everything is in good shape and this will make for a great addition to the agent capabilities.

The code all looks good. I've reviewed the latest changes.

I've tested the capabilities with agents and CLI. Things behave fine. My testing has been mostly happy-path. I've tested out some conflicting or incorrect parameters but I've only tested on a reliable network with straightforward configurations and use. I haven't done any robustness testing.

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
Merge branch 'master' of github.com:jenkinsci/jenkins into websocket

jglick committed:
https://github.com/jenkinsci/remoting/pull/357 released as 4.0.

oleg-nenashev commented:
@jglick what is the rollout plan for this change? Do you want to have it as `Beta` in the next LTS or just as GA? It might be useful to put it into the JEP explicitly

fcojfernandez reviewed (approved):
I haven't tested it, but the code looks good

jglick commented:
> Do you want to have it as `Beta` in the next LTS or just as GA?

I do not think there is any formal mechanism for any such marking. We can phrase it as being experimental in the weekly changelog of course.

oleg-nenashev commented:
> I do not think there is any formal mechanism for any such marking. We can phrase it as being experimental in the weekly changelog of course.

Changelog would be enough. APIs are already marked as `@Beta`

jglick commented:
> APIs are already marked as `@Beta`

And could probably stay that way unless plugins start to want to implement their own endpoints.

timja reviewed (approved):
Haven't tested but code looks good

oleg-nenashev commented:
requesting reviews from the @jenkinsci/core team since it is a major change under the hood of Jenkins. Unfortunately I do not have bandwidth to review it till the next week (wed or so), but I do not mind if it gets merged.with the current approvals.

@jglick Could you please also add the "collateral improvements" from Remoting and Winstone to the changelog?

* Major RFE: Remoting 4.0 release, mostly to highlight the binary incompatible in 4.0 and previous 3.x. https://github.com/jenkinsci/remoting/releases/tag/remoting-4.0
* RFE: Jetty update in Winstone 5.5: https://github.com/jenkinsci/winstone/releases/tag/winstone-5.5

oleg-nenashev commented:
I plan to merge it tomorrow if no negative feedback

jglick commented:
> requesting reviews from the @jenkinsci/core team

Note that I already requested reviews from @jenkinsci/core-pr-reviewers.

oleg-nenashev commented:
@jglick would you be interested to write a blog post about this new feature?

jglick commented:
Yes I should be able to do that.

jglick commented:
Merged toward 2.216 FTR.
